source/common/http/alternate_protocols.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "envoy/common/time.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Http {
+
+// Tracks alternate protocols that can be used to make an HTTP connection to an origin server.
+class AlternateProtocols {
+public:
+  // Represents an HTTP origin to be connected too.
+  struct Origin {
+  public:
+    Origin(absl::string_view scheme, absl::string_view hostname, int port);
+
+    bool operator==(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) ==
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator!=(const Origin& other) const { return !this->operator==(other); }
+
+    bool operator<(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) <
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator>(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) >
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator<=(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) <=
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    bool operator>=(const Origin& other) const {
+      return std::tie(scheme_, hostname_, port_) >=
+             std::tie(other.scheme_, other.hostname_, other.port_);
+    }
+
+    std::string scheme_;
+    std::string hostname_;
+    int port_{};
+  };
+
+  // Represents an alternative protocol that can be used to connect to an origin.
+  struct AlternateProtocol {
+  public:
+    AlternateProtocol(absl::string_view alpn, absl::string_view hostname, int port);
+
+    bool operator==(const AlternateProtocol& other) const {
+      return std::tie(alpn_, hostname_, port_) ==
+             std::tie(other.alpn_, other.hostname_, other.port_);
+    }
+
+    bool operator!=(const AlternateProtocol& other) const { return !this->operator==(other); }
+
+    std::string alpn_;
+    std::string hostname_;
+    int port_;
+  };
+
+  explicit AlternateProtocols(TimeSource& time_source);
+
+  // Sets the possible alternative protocols which can be used to connect to the
+  // specified origin. Expires after the specified expiration time.
+  void setAlternatives(const Origin& origin, const std::vector<AlternateProtocol>& protocols,
+                       const MonotonicTime& expiration);
+
+  // Returns the possible alternative protocols which can be used to connect to the
+  // specified origin, or nullptr if not alternatives are found. The returned pointer
+  // is owned by the AlternateProtocols and is valid until the next operation on
+  // AlternateProtocols.
+  const std::vector<AlternateProtocol>* findAlternatives(const Origin& origin);
```

Comment:
Arguably it's a little more Envoy-ish to return an absl::optional<std::reference_wrapper<<>> but honestly I find pointers way easier to read so I'm going to look the other way here as long as you make the return value const :-P

Reply:
some of the messiness of opt ref is encapsulated in Envoy::OptRef : https://github.com/envoyproxy/envoy/blob/main/include/envoy/common/optref.h

In case that helps.

Reply:
Ooh, interesting! I've switched to OptRef, thanks!

Does Optref provide some additional type safety beyond a naked pointer?

Reply:
Not particularly; it just perhaps conveys intent a little better rather than a null check, and, with the convenience methods on that class, is IMO more on par from a readability perspective.